# Trypanosoma_cruzi-host-microbiome
# Microbiome alterations driven by Trypanosoma cruzi infection in two disjunctive mouse models

Alterations caused by Trypanosoma cruzi in the gut microbiome composition may play a key role in the host-parasite interactions and may be involved in a decrease of bacterial taxa and metabolic pathways that could be related to changes in physiology and immune responses against the parasite, promoting the establishment and progression of the infection. Therefore, we implemented a murine model with two mouse strains, BALBc and BL6, to evaluate the impact of Trypanosoma cruzi (Tulahuen strain) infection on the gut microbiome using shotgun metagenomics.

Bioinformatics Analysis

From the raw reads, a quality assessment was performed by FastQC (25). Quality and adapter trimming was then performed by Trimmomatic (26) using the parameters ILLUMINACLIP: TruSeq3-PE.fa:2:30:10:2:keepBothReads MINLEN:150 AVGQUAL:20 TRAILING:20. Mapping to the Mus musculus genome (GRCm39) was performed using the Bowtie2 tool (27) to remove reads corresponding to the host.

From the clean reads, the Centrifuge tool was used for taxonomic assignment (28). The obtained outputs were transformed to Kraken-Report format with the Centrifuge-kreport function for the corresponding analysis and visualization by Pavian (29). Simultaneously, Humann3 (HMP Unified Metabolic Analysis Network) (30) was used for functional profiling including the identification of changes in gene abundance and metabolic pathways. Since this tool receives only one fastq file as input, forward and reverse reads from each sample were concatenated to generate such a file. To facilitate comparisons between samples with different sequencing depths, it is important to normalize RPK values to relative abundance values or "copies per million" (CPM) units, so the humann_renorm_tablescript function was used. The default "units" of the HUMAnN microbial function are the UniRef gene families that were used to calculate reaction and pathway abundances. From the abundance of gene families, reconstruction of the abundance of other functional categories was performed using the humann_regroup_tablescript function. Gene family abundance values normalized by CPM to the MetaCyc reaction abundances (RXN), which are included with the default Humann3 facility, were regrouped. To generate readable output, the humann_rename_table function was used, and then the sample outputs were joined with the humann_join_tables function. Finally, a stratified table was obtained with humann_split_stratified_table. For the evaluation of the results obtained by Humann3, the R package, Maaslin (31), was used to determine the association between measurement times, condition (infected-uninfected,) and functional characteristics of the microbiome using a multivariate mixed-effects model.
